### ISSUE TITLE
Numbers in scientific notation are not supported

### ISSUE THREAD

**Issue by euglv** [Pomax/svg-to-something#6]
For example, when doing reverse of the following path: 
`M 1200 0 A 1200 1200 0 0 1 848.5281374238571 848.528137423857 A 1200 1200 0 0 1 7.347880794884119e-14 1200`  
The result is: 
`M 7.347880794884119 undefined A 1200 1200 0 0 0 848.5281374238571 848.528137423857 A 1200 1200 0 0 0 1200 0` 

Numbers `7.347880794884119e-14 1200` are treated as `7.347880794884119 undefined` instead of `0.00000000000007347880794884119 1200`

**Comment by euglv** [comment]
2 changes are required to fix this issue:
Change 25 line from:
```js
.replace(/-/g,' -')
```
To:
```js
.replace(/(?<![eE])-/g,' -')
```
But negative look behind may be not supported by some browsers, may be there is a better fix for this line.

And change 30 line from:
```js
var instructions = d.replace(/([a-zA-Z])\s?/g,"|$1").split("|"),
```
To:
```js
var instructions = d.replace(/([achlmqstvzACHLMQSTVZ])\s?/g,"|$1").split("|"),
```

**Comment by Pomax** [comment]
I'd rather not, for a very simple reason: the SVG unit of measurement is the "unitless unit", e.g. you're working on a grid where sensible numbers are integer aligned, and you might have a few decimals, and you might have a few non-decimal digits, but you're never going to have numbers so big, or so small, that scientific notation is required to represent them.

For example: your SVG shows `7.347880794884119e-14`, but what that's telling you is that you have bugs that need to be looked at, either in code, or in process: that's the number zero, not the number 0.0000000000000073, and you should be fixing whatever's not properly generating the values you need. The SVG scale is a few orders of magnitude, four is not unusual, five is pushing it, but we're seeing EIGHTEEN orders of magnitude between the largest and smallest values. That's ludicrous: you need to fix something here, because something's very broken.

If this tool can surface that problem for you, I'll take it. At least now you know your SVG has bugs.

**Comment by euglv** [comment]
No, SVG unit of measurement is not always the "unitless unit". 
https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Positions
> In the most basic case one pixel in an SVG document maps to one pixel on the output device (a.k.a. the screen). But SVG wouldn't have the "Scalable" in its name, if there weren't several possibilities to change this behavior. Much like absolute and relative font sizes in CSS, **SVG defines absolute units (ones with a dimensional identifier like "pt" or "cm")** and so-called user units, that lack that identifier and are plain numbers.

SVGs are not always meant to be rendered on the grid where sensible numbers are integer aligned. Some one may want to use SVG to "print" it on CNC Laser Cutting Machine.
Can you point where SVG specification states that numbers should be rounded? And if they should be rounded, then to what precision?
I can point that SVG specification states that scientific notation is completely valid in SVG:
https://lists.w3.org/Archives/Public/www-svg/1999Nov/0027.html

> If anyone finds an implementation which fails on floating point
> coordinates, please mail the maintainer of that software and refer them
> to this URL:
> 
> http://www.w3.org/1999/08/WD-SVG-19990812/paths.html#PathDataBNF

I understand, that in most cases there would be no noticeable difference between `7.347880794884119e-14` and `0`.
But I am generating SVG this way:
```js
path += ` M ${x} ${y}`
```
If I add rounding everywhere in my code then it would become too ugly, and not simple and harder to read. And I do not care about SVG size and how ugly will SVG be, as no one is going to read the generated SVGs.

**Comment by Pomax** [comment]
> No, SVG unit of measurement is not always the "unitless unit".

SVG _path values_, which is the only thing this library cares about, are.

> SVGs are not always meant to be rendered on the grid where sensible numbers are integer aligned. Some one may want to use SVG to "print" it on CNC Laser Cutting Machine.

CnC machines don't care about numbers in the e-14 range. They care about numbers up to their precision (usually measured in the 1-0.1mm for metric / 10-1 thou for imperial range) so if you're going to use it for a laser cutter or a water jet, fix your numbers to make sense for that device. Feeding it e-14 numbers is literally meaningless.

>Can you point where SVG specification states that numbers should be rounded? And if they should be rounded, then to what precision?

Depends: are you designing it for a screen? 1 decimal is plenty. Are you designing it for an hobbyist CnC device? 2 decimals is plenty (getting you 0.1mm/ 4 thou resolution). Professional machining? 3 decimals is plenty (0.01mm/0.4 thou resolution).

> If I add rounding everywhere in my code then it would become too ugly

I didn't say to round your numbers to integers: round your numbers to _the precision you need_. In this case, your largest number is 1200, so rounding to one decimal is sensible, and rounding to two decimals could be necessary, or could be overkill, and probably the latter - but, we're not even talking about rounding here, we're talking about a number _that is zero_: just make it zero. No technology exists that can meaningfully deal with a value that small, you'd need literally a million million (that's 1000000000000) servo backlashes before you'd even start to notice being off by an almost significant value. There's not enough time in the universe for that.

So again, you will never need scientific numbers in your SVG, because even though they're "valid", there's no meaning to them. If you do see them, you should ask why, and fix that. And if you don't want to fix it manually, run your SVG through an optimizer first, and set that to automatically fix crazy numbers like these.

So in conclusion: are they valid? Yes. Are they _sensible_? Never. If you can show me an _actually sensible_ SVG where scientific notation is not just an over-engineered standin for zero in the `<path d="..."`>` attribute, by all means prove me wrong, but just because it's supported, doesn't mean it makes sense, and doesn't mean you should want it to work when you have ridiculous values in your pathing instructions.

**Comment by euglv** [comment]
Even if I work with screen, rounding to some precision is only needed to save memory. But code that do a lot of rounding in a loop is less performant. And, depending on the situation, some times it is better to have code as much simple as possible than to consume a little bit less memory.

Here is example of code that generates regular polygon with _n_ vertices inscribed into a circle of radius _r_:
```js
function regularPolygon(centerX, centerY, r, n){
    let path = `M ${centerX + r} ${centerY} `;
    let step = 2 * Math.PI / n;

    for (let i = 1; i < n; i++) {
        let angle = i * step;
        let x = r * Math.cos(angle);
        let y = r * Math.sin(angle);
        path += `L ${x} ${y} `;
    }
    path += 'Z';
    return path;
}
```
So what's wrong with the above function? What bugs does it have? It always produces a valid SVG path according to SVG specification. It is not doing rounding to some precision for the sake of simplicity and performance.
But in a very rare cases it can produce "zeros" in scientific notation:
```js
console.log(regularPolygon(0, 0, 100, 3)); 
//Otputs: M 100 0 L -49.99999999999998 86.60254037844388 L -50.00000000000004 -86.60254037844385 Z

console.log(regularPolygon(0, 0, 100, 4)); 
//Otputs: M 100 0 L 6.123233995736766e-15 100 L -100 1.2246467991473532e-14 L -1.8369701987210297e-14 -100 Z

console.log(regularPolygon(0, 0, 100, 5)); 
//Otputs: M 100 0 L 30.901699437494745 95.10565162951535 L -80.90169943749473 58.77852522924732 L -80.90169943749474 -58.7785252292473 L 30.901699437494724 -95.10565162951536 Z
```
But in the example above, if you move the center of the polygon to the point `(150;150)` - you will never get "zeros" in scientific notation.

**Comment by Pomax** [comment]
And there, too: run your SVG through an optimizer first then. Let's see what [svgo](https://github.com/svg/svgo) does (which you can run universally either by installing globally, or by using `npx svgo`, in which case you don't even need to have it persistently installed):

```xml
<svg>
  <path d="M 100 7.14e-14 L -49.99999999999998 86.60254037844388 L -50.00000000000004 -86.60254037844385 Z"/>
</svg>
```

becomes

```xml
<svg><path d="M100 0-50 86.603V-86.603z"/></svg>
```

Perfect: just use an SVG optimizer. Heck, not even related to this particular problem: _always_ use an SVG optimizer. 

And if you're only generating pathing, never generating actual full SVG documents (in which case almost all the complaints above are irrelevant because you're clearly _not_ using it as SVG document for some fabrication or screen purpose), then just add a few lines of code for injecting your coordinate values using a `coord` function that makes sure your numbers have the required resolution:

```js
const { sin, cos, PI } = Math;

function coord(x, y, maxDecimals=3) {
  x = parseFloat(x.toFixed(maxDecimals)); // if you've never seen this trick: it removes trailing zeroes
  y = parseFloat(y.toFixed(maxDecimals));
  return `${x} ${y}`;
}

function regularPolygon(centerX, centerY, r, n){
    let path = `M ${coord(centerX + r, centerY)}`;
    let step = 2 * PI / n;

    for (let i = 1; i < n; i++) {
        let angle = i * step;
        let x = r * cos(angle);
        let y = r * sin(angle);
        path += `L ${coord(x,y)}`;
    }

    path += 'Z';
    return path;
}
```

So let's run that:

```
>> console.log(regularPolygon(0, 0, 100, 5));
M 100 0L 30.902 95.106L -80.902 58.779L -80.902 -58.779L 30.902 -95.106Z
```

Again, perfect. 

Nothing complicated, just write your code for the resolution your path needs to be, and get a clean path. The tiniest bit of work up front gets you incredibly better markup forever after. That's worth doing.
